doc/languages-frameworks/java.xml

```diff
@@ -43,7 +43,21 @@ installPhase =
       --add-flags "-cp $out/share/java/foo.jar org.foo.Main"
   '';
 </programlisting>
-  Note the use of <literal>jre</literal>, which is the part of the OpenJDK package that contains the Java Runtime Environment. By using <literal>${jre}/bin/java</literal> instead of <literal>${jdk}/bin/java</literal>, you prevent your package from depending on the JDK at runtime.
+  Since the introduction of the 'modules' system in Java 9, distributions typically no longer ship with a general-purpose JRE: instead, they allow generating a JRE with only the modules required for your application(s). Because we can't predict what modules will be needed on a general-purpose system, the default <literal>jre</literal> is the full JDK. When building a minimal system/image, you can use the <literal>jre_minimal</literal> derivation to build a JRE with only the modules relevant for you:
```

Comment:
adopted the first suggestion, further improved the wording on the second half

Reply:
The way you've phrased it assumes you're configuring a global JRE for a "system" (as in NixOS, presumably). Most usages of JREs will be in package definitions though, which is why I abstracted the use case of the JRE in my rephrasing.

Reply:
That was sort of intentional.

With "system" I meant any system/image/... built with Nix (NixOS or otherwise). Packages that require a JRE should by default depend on `jre`, not on `jre_minimal`. Only when you want to minimize your configuration you would use `jre_minimal` to build a JRE with only the modules you need, and override the default `jre` with that custom one. You would never override `jre_minimal` because no package would depend on jre_minimal by default.

The reason for this is that say when 'something' depends on a JRE with modules A, B and C, and 'other' depends on a JRE with modules C and D, you don't want to install a JRE with ABC and a JRE with CD, but one JRE with ABCDE. Since individual packages can't predict what they'll be installed next to, the default jre would still be a 'full', non-minimal JRE.

You could imagine using `jre_minimal` with all modules enabled as the default jre. While I see the appeal, that is different than 'common' and it is hard to predict what kinds of problems that might introduce (I could imagine it behaves differently with respect to `sun.misc` usage, for example), so I think that would be too dangerous/invasive as a default.

Reply:
I guess it's the best we can do now.